File: host-rs/src/device_readiness.rs
//! Device readiness parsing, mirroring the C# `DeviceReadiness`.

#[derive(Debug, Clone)]
pub struct DeviceReadiness {
    pub ready: bool,
    pub reason: String,
}

impl DeviceReadiness {
    pub fn parse(output: &str) -> Self {
        let trimmed = output.trim();
        if trimmed.is_empty() {
            return Self { ready: false, reason: "No response from device".into() };
        }
        if trimmed == "READY" {
            return Self { ready: true, reason: "Ready (unlocked)".into() };
        }
        if trimmed.eq_ignore_ascii_case("SCREEN_OFF") {
            return Self { ready: false, reason: "Screen is off / asleep".into() };
        }
        if trimmed.eq_ignore_ascii_case("LOCKED") {
            return Self { ready: false, reason: "Device is locked".into() };
        }
        Self { ready: false, reason: format!("Device not ready ({trimmed})") }
    }

    /// True when the failure is just adb being slow/timing out, not a real device state.
    pub fn is_adb_transient(&self) -> bool {
        let r = self.reason.to_ascii_lowercase();
        r.contains("adb shell command failed") || r.contains("readiness check failed")
    }
}

File: host-c#/SecondDisplay.Host/DeviceReadiness.cs
using System;

namespace SecondDisplay.Host;

/// <summary>
/// Status check indicating whether the tablet is unlocked, awake, and in "data transfer" (MTP/PTP) USB mode.
/// </summary>
public readonly record struct DeviceReadiness(bool IsReady, string Reason)
{
    public static DeviceReadiness Parse(string? output)
    {
        if (string.IsNullOrWhiteSpace(output))
            return new DeviceReadiness(false, "No response from device");

        string trimmed = output.Trim();
        if (trimmed == "READY")
            return new DeviceReadiness(true, "Ready (unlocked, MTP active)");

        if (trimmed.StartsWith("USB_NOT_MTP", StringComparison.OrdinalIgnoreCase))
        {
            int colon = trimmed.IndexOf(':');
            string state = colon >= 0 ? trimmed[(colon + 1)..].Trim() : "";
            string reason = string.IsNullOrEmpty(state)
                ? "USB mode is not data transfer"
                : $"USB mode is not data transfer ({state})";
            return new DeviceReadiness(false, reason);
        }

        if (trimmed == "LOCKED")
            return new DeviceReadiness(false, "Device is locked");

        if (trimmed == "SCREEN_OFF")
            return new DeviceReadiness(false, "Screen is off / asleep");

        return new DeviceReadiness(false, $"Device not ready ({trimmed})");
    }
}
